raven/src/test/java/org/opencadc/raven/ProtocolsGeneratorTest.java

```diff
@@ -100,5 +107,141 @@ public void testPrioritizePullFromSites() throws Exception {
             Assert.assertFalse(!sites.get(i - 1).getAllowWrite() && sites.get(i).getAllowWrite());
         }
     }
+
+    @Test
+    public void testPrioritizingStorageSiteComparator() throws Exception {
+
+        Map<URI, StorageSiteRule> siteRules = new HashMap<>();
+        InetAddress clientIP = null;
+
+        ProtocolsGenerator.PrioritizingStorageSiteComparator comparator =
+            new ProtocolsGenerator.PrioritizingStorageSiteComparator(siteRules, URI.create("ivo:aaa"), clientIP);
+
+        // both StorageSite's null
+        int actual = comparator.compare(null, null);
+        Assert.assertEquals(0, actual);
+
+        //2nd StorageSite null
+        actual = comparator.compare(new StorageSite(URI.create("ivo:site1"), "site1", true, true), null);
+        Assert.assertEquals(-1, actual);
+
+        // 1st StorageSite null
+        actual = comparator.compare(null, new StorageSite(URI.create("ivo:site1"), "site1", true, true));
+        Assert.assertEquals(1, actual);
+
+
```

Comment:
Here you should have a compare test for the same two sites (2 and 5) and no rules so that you know it falls through to the StorageSite natural order and confirm the expected natural order. Otherwise, your test with a rule to "swap" the order below aren't necessarily working as intended.

In theory StorageSite could evolve and natural order could change, but since it is currently defined to use the resourceID that's a globally unique ID it seems unlikely. But: I still like to have a test that verifies the test conditions (it's not really an assumption). I wouldn't go so far as to make the test resilient to natural order changes because it's so unlikely and would obfuscate the test.

Reply:
Added another test to with no site rules to confirm the default StorageSite order is used.

Not understanding the second paragraph, mainly `a test that verifies the test conditions`?

Reply:
In the case where you expect the rule to change the order, you need  the non-rule test first to make sure that isn't already the order; for example, if you had added the preference to site2 it would not change the order and your test wouldn't prove that rules change the order in the right way. 

So: one compare/assert for site2 vs site5 that verifies that site2 < site5 by default and then the real test w/ rule that makes site5 < site2. That lessens the chance that a bug goes undetected because one of those will fail if it isn't working as intended.

That means the asserts on line 147 and 152 *do care* whether the result is -1 or 1; all the others are OK.